src/main/java/org/mitre/hapifhir/MedMorphToCIBMTR.java

```diff
@@ -1,58 +1,63 @@
 package org.mitre.hapifhir;
 
-import java.util.List;
-import java.util.Map;
-import java.net.URL;
 import java.net.HttpURLConnection;
 import java.net.MalformedURLException;
-import org.hl7.fhir.r4.model.Bundle;
-import org.hl7.fhir.r4.model.Bundle.BundleEntryComponent;
-import org.hl7.fhir.r4.model.Resource;
-import org.hl7.fhir.r4.model.ResourceType;
-import org.hl7.fhir.r4.model.Patient;
-import org.hl7.fhir.r4.model.Observation;
-import org.hl7.fhir.r4.model.CodeableConcept;
-import org.hl7.fhir.r4.model.Coding;
-import org.hl7.fhir.r4.model.Quantity;
-import org.json.JSONObject;
-import org.json.JSONArray;
+import java.net.URL;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
 import org.apache.http.HttpEntity;
-import org.apache.http.client.methods.HttpPut;
 import org.apache.http.client.methods.HttpPost;
+import org.apache.http.client.methods.HttpPut;
 import org.apache.http.client.ResponseHandler;
 import org.apache.http.entity.StringEntity;
 import org.apache.http.impl.client.CloseableHttpClient;
 import org.apache.http.impl.client.HttpClients;
 import org.apache.http.util.EntityUtils;
-import java.util.stream.Collectors;
+import org.hl7.fhir.r4.model.Bundle;
+import org.hl7.fhir.r4.model.Bundle.BundleEntryComponent;
+import org.hl7.fhir.r4.model.CodeableConcept;
+import org.hl7.fhir.r4.model.Coding;
+import org.hl7.fhir.r4.model.Identifier;
+import org.hl7.fhir.r4.model.MessageHeader;
+import org.hl7.fhir.r4.model.Observation;
+import org.hl7.fhir.r4.model.Organization;
+import org.hl7.fhir.r4.model.Patient;
+import org.hl7.fhir.r4.model.Quantity;
+import org.hl7.fhir.r4.model.Reference;
+import org.hl7.fhir.r4.model.Resource;
+import org.hl7.fhir.r4.model.ResourceType;
+import org.json.JSONArray;
+import org.json.JSONObject;
 
 public class MedMorphToCIBMTR {
+  private static final String CCN_SYSTEM = "http://cibmtr.org/codesystem/transplant-center";
+  private static final String CRID_SYSTEM = "http://cibmtr.org/identifier/CRID";
   private String cibmtrUrl;
-  private String ccn;
 
-  public MedMorphToCIBMTR(String cibmtrUrl, String ccn) {
+  public MedMorphToCIBMTR(String cibmtrUrl) {
     this.cibmtrUrl = cibmtrUrl;
     if (!this.cibmtrUrl.endsWith("/")) this.cibmtrUrl += "/";
-    this.ccn = ccn;
   }
 
-  public void convert(Bundle medmorphReport, String authToken) {
+  public void convert(Bundle medmorphReport, MessageHeader messageHeader, String authToken) {
     // https://fhir.nmdp.org/ig/cibmtr-reporting/CIBMTR_Direct_FHIR_API_Connection_Guide_STU3.pdf
     if (medmorphReport.hasEntry()) {
       List<BundleEntryComponent> entriesList = medmorphReport.getEntry();
       BundleEntryComponent patientEntry = entriesList.stream().filter(entry -> entry.getResource().getResourceType() == ResourceType.Patient).findAny().orElse(null);
```

Comment:
This may have been the bug you mentioned you noticed this morning, but finding the patientEntry here and passing the entries list below when we post the bundle is missing a piece. The MedMorph report format is actually "double wrapped" in the sense that there's an outer Bundle, containing the MessageHeader, content Bundle, and Organization, and then the content Bundle is what contains the real data, such as the Patient, Condition, Observations, etc. I think this just needs another line or 2 to pick out the content bundle and use that to find the patient and observations. There are 2 ways to identify the content bundle, 1 - it's the `focus` of the MessageHeader, and 2 it should probably be the second resource in the report bundle.

Other than that everything looks pretty good.

Reply:
Ahh I see, just pushed a commit that should deal with the double wrapped bundle.